File: py-dfe/py_dfe/xmlops/processor.py
"""Build processed DFe XML documents (nfeProc, cteProc, mdfeProc, procEvento*)."""

from __future__ import annotations

import copy
import logging
from typing import Any

from lxml import etree

logger = logging.getLogger(__name__)

_NS: dict[str, str] = {
    "nfe": "http://www.portalfiscal.inf.br/nfe",
    "nfce": "http://www.portalfiscal.inf.br/nfe",
    "cte": "http://www.portalfiscal.inf.br/cte",
    "mdfe": "http://www.portalfiscal.inf.br/mdfe",
}

# service_name: (proc_root_tag, doc_local_tag, prot_local_tag, versao)
_EMISSION: dict[str, tuple[str, str, str, str]] = {
    "NFeAutorizacao": ("nfeProc", "NFe", "protNFe", "4.00"),
    "NfceAutorizacao": ("nfeProc", "NFe", "protNFe", "4.00"),
    "CTeRecepcaoSinc": ("cteProc", "CTe", "protCTe", "4.00"),
    "CTeRecepcaoOS": ("cteOSProc", "CTeOS", "protCTe", "4.00"),
    "CTeRecepcaoGTVe": ("GTVeProc", "GTVe", "protCTe", "4.00"),
    "CTeRecepcaoSimp": ("cteSimpProc", "CTeSimp", "protCTe", "4.00"),
    "MDFeRecepcaoSinc": ("mdfeProc", "MDFe", "protMDFe", "3.00"),
}

# service_name: (proc_root_tag, versao)
# event elements are always renamed to 'evento'/'retEvento' in the proc document
_EVENT: dict[str, tuple[str, str]] = {
    "RecepcaoEvento": ("procEventoNFe", "1.00", "evento", "retEvento"),
    "CTeRecepcaoEvento": ("procEventoCTe", "4.00" "eventoCTe", "retEventoCTe"),
    "MDFeRecepcaoEvento": ("procEventoMDFe", "3.00", "eventoMDFe", "retEventoMDFe"),
}


def build_processed_xml(
        doc_type: str,
        service: str,
        request_xml: bytes,
        response_xml: bytes,
) -> str | None:
    """Return the processed XML string, or None if the service has no processed form."""
    try:
        ns = _NS.get(doc_type)
        if not ns:
            return None
        if service in _EMISSION:
            return _build_emission(ns, service, request_xml, response_xml)
        if service in _EVENT:
            return _build_events(ns, service, request_xml, response_xml)
    except Exception:
        logger.debug(
            "Could not build processed XML for %s/%s", doc_type, service, exc_info=True
        )
    return None


# ---------------------------------------------------------------------------
# Emission: nfeProc, cteProc, mdfeProc, etc.
# ---------------------------------------------------------------------------


def _build_emission(
        ns: str,
        service: str,
        request_xml: bytes,
        response_xml: bytes,
) -> str | None:
    proc_tag, doc_tag, prot_tag, versao = _EMISSION[service]

    req_root = etree.fromstring(request_xml)
    resp_root = etree.fromstring(response_xml)

    doc_el = _first_by_local(req_root, doc_tag)
    prot_el = _first_by_local(resp_root, prot_tag)

    if doc_el is None or prot_el is None:
        return None

    proc = etree.Element(f"{{{ns}}}{proc_tag}", nsmap={None: ns})
    proc.set("versao", versao)
    proc.append(_deep_copy(doc_el))
    proc.append(_deep_copy(prot_el))

    return etree.tostring(proc, encoding="unicode", xml_declaration=False)


# ---------------------------------------------------------------------------
# Events: procEventoNFe, procEventoCTe, procEventoMDFe
# ---------------------------------------------------------------------------


def _build_events(
        ns: str,
        service: str,
        request_xml: bytes,
        response_xml: bytes,
) -> str | None:
    proc_tag, versao, event_tag, ret_event_tag = _EVENT[service]

    req_root = etree.fromstring(request_xml)
    resp_root = etree.fromstring(response_xml)

    # Events in request may be called 'evento', 'eventoCTe', 'eventoMDFe', etc.
    events = _all_by_local(req_root, event_tag)
    if not events:
        # fallback: whole request root IS the event (single event, no envelope)
        events = [req_root]

    # Response may wrap under retEnvEvento/retEvento or retEvento directly
    ret_events = _all_by_local(resp_root, ret_event_tag)
    if not ret_events:
        return None

    results: list[str] = []
    for ev, ret_ev in zip(events, ret_events):
        proc = etree.Element(f"{{{ns}}}{proc_tag}", nsmap={None: ns})
        proc.set("versao", versao)
        # Always tag child as 'evento' (rename if needed)
        proc.append(_as_local_tag(ev, "evento", ns))
        # Always tag child as 'retEvento' (rename if needed)
        proc.append(_as_local_tag(ret_ev, "retEvento", ns))
        results.append(etree.tostring(proc, encoding="unicode", xml_declaration=False))

    if not results:
        return None
    if len(results) == 1:
        return results[0]
    # Multiple events: JSON list of XML strings
    import json

    return json.dumps(results)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _all_by_local(root: etree.Element, local_tag: str) -> list[etree.Element]:
    """Return all descendants (including root) with the given local tag."""
    if etree.QName(root.tag).localname == local_tag:
        return [root]
    return [el for el in root.iter() if etree.QName(el.tag).localname == local_tag]


def _first_by_local(root: etree.Element, local_tag: str) -> etree.Element | None:
    els = _all_by_local(root, local_tag)
    return els[0] if els else None


def _deep_copy(el: etree.Element) -> etree.Element:
    return copy.deepcopy(el)


def _as_local_tag(el: etree.Element, local_tag: str, ns: str) -> etree.Element:
    """Return a deep copy of el, with its root tag renamed to local_tag in ns."""
    new_el = copy.deepcopy(el)
    if etree.QName(new_el.tag).localname != local_tag:
        new_el.tag = f"{{{ns}}}{local_tag}"
    return new_el

File: go-dfe/internal/xmlops/processor.go
package xmlops

import "strings"

// emissionSpec/eventSpec mirror py-dfe's _EMISSION/_EVENT tuples
// (py-dfe/py_dfe/xmlops/processor.py).
type emissionSpec struct{ procTag, docTag, protTag, versao string }
type eventSpec struct{ procTag, versao, eventTag, retEventTag string }

// emissionServices mirrors py-dfe's _EMISSION.
var emissionServices = map[string]emissionSpec{
	"NFeAutorizacao":   {"nfeProc", "NFe", "protNFe", "4.00"},
	"NfceAutorizacao":  {"nfeProc", "NFe", "protNFe", "4.00"},
	"CTeRecepcaoSinc":  {"cteProc", "CTe", "protCTe", "4.00"},
	"CTeRecepcaoOS":    {"cteOSProc", "CTeOS", "protCTe", "4.00"},
	"CTeRecepcaoGTVe":  {"GTVeProc", "GTVe", "protCTe", "4.00"},
	"CTeRecepcaoSimp":  {"cteSimpProc", "CTeSimp", "protCTe", "4.00"},
	"MDFeRecepcaoSinc": {"mdfeProc", "MDFe", "protMDFe", "3.00"},

	// Inutilização não é emissão, mas tem exatamente a mesma forma processada:
	// documento assinado + resposta da SEFAZ sob uma raiz única. O elemento é
	// ProcInutNFe (P maiúsculo, procInutNFe_v4.00.xsd), e o "documento" é a
	// própria raiz do request — firstByLocal inclui a raiz na busca.
	"NfeInutilizacao": {"ProcInutNFe", "inutNFe", "retInutNFe", "4.00"},
}

// eventServices mirrors py-dfe's _EVENT. py-dfe's own CTeRecepcaoEvento
// entry has a latent bug there: adjacent string literals ("4.00" "eventoCTe")
// concatenate in Python, collapsing what should be a 4-tuple into 3 elements
// — the unpack (`proc_tag, versao, event_tag, ret_event_tag = _EVENT[service]`)
// raises at runtime, which build_processed_xml's blanket try/except silently
// swallows (logs at debug, returns None). CTeRecepcaoEvento functionally
// never gets a processed XML in py-dfe today because of this. Ported here
// with the obviously-intended 4 values instead of reproducing the crash —
// matching an accidental typo helps no one — flagged here for awareness.
var eventServices = map[string]eventSpec{
	"RecepcaoEvento":     {"procEventoNFe", "1.00", "evento", "retEvento"},
	"CTeRecepcaoEvento":  {"procEventoCTe", "4.00", "eventoCTe", "retEventoCTe"},
	"MDFeRecepcaoEvento": {"procEventoMDFe", "3.00", "eventoMDFe", "retEventoMDFe"},
}

// docNamespaces mirrors processor.py's _NS.
var docNamespaces = map[string]string{
	"nfe":  "http://www.portalfiscal.inf.br/nfe",
	"nfce": "http://www.portalfiscal.inf.br/nfe",
	"cte":  "http://www.portalfiscal.inf.br/cte",
	"mdfe": "http://www.portalfiscal.inf.br/mdfe",
}

// BuildProcessedXML returns the processed XML string (nfeProc/cteProc/
// mdfeProc/procEventoNFe/etc — the signed request document and SEFAZ's
// protocol/event response combined into the single document SEFAZ
// convention expects for storage), or ("", false) if service has no
// processed form. Mirrors py-dfe's build_processed_xml
// (py-dfe/py_dfe/xmlops/processor.py), including its behavior of silently
// returning false on any malformed input — a processed-XML failure must
// never break the underlying SEFAZ call it's attached to.
func BuildProcessedXML(docType, service string, requestXML, responseXML []byte) (result string, ok bool) {
	defer func() {
		if r := recover(); r != nil {
			result, ok = "", false
		}
	}()

	ns, hasNS := docNamespaces[docType]
	if !hasNS {
		return "", false
	}
	if spec, isEmission := emissionServices[service]; isEmission {
		return buildEmission(ns, spec, requestXML, responseXML)
	}
	if spec, isEvent := eventServices[service]; isEvent {
		return buildEvents(ns, spec, requestXML, responseXML)
	}
	return "", false
}

func buildEmission(ns string, spec emissionSpec, requestXML, responseXML []byte) (string, bool) {
	reqRoot, respRoot, err := parseRoots(requestXML, responseXML)
	if err != nil {
		return "", false
	}

	docEl := firstByLocal(reqRoot, spec.docTag)
	protEl := firstByLocal(respRoot, spec.protTag)
	if docEl == nil || protEl == nil {
		return "", false
	}

	proc := newElem(spec.procTag)
	proc.nsDecls = []nsDecl{{prefix: "", uri: ns}}
	proc.attrs = []attrNode{{local: "versao", value: spec.versao}}
	proc.appendChild(deepCopyNode(docEl))
	proc.appendChild(deepCopyNode(protEl))

	return serializeStandalone(proc), true
}

func buildEvents(ns string, spec eventSpec, requestXML, responseXML []byte) (string, bool) {
	reqRoot, respRoot, err := parseRoots(requestXML, responseXML)
	if err != nil {
		return "", false
	}

	events := allByLocal(reqRoot, spec.eventTag)
	if len(events) == 0 {
		events = []*xNode{reqRoot}
	}
	retEvents := allByLocal(respRoot, spec.retEventTag)
	if len(retEvents) == 0 {
		return "", false
	}

	n := len(events)
	if len(retEvents) < n {
		n = len(retEvents)
	}
	results := make([]string, 0, n)
	for i := 0; i < n; i++ {
		proc := newElem(spec.procTag)
		proc.nsDecls = []nsDecl{{prefix: "", uri: ns}}
		proc.attrs = []attrNode{{local: "versao", value: spec.versao}}
		proc.appendChild(asLocalTag(events[i], "evento", ns))
		proc.appendChild(asLocalTag(retEvents[i], "retEvento", ns))
		results = append(results, serializeStandalone(proc))
	}
	if len(results) == 0 {
		return "", false
	}
	if len(results) == 1 {
		return results[0], true
	}
	// Multiple events: py-dfe returns a JSON array of XML strings.
	return jsonStringArray(results), true
}

func parseRoots(requestXML, responseXML []byte) (reqRoot, respRoot *xNode, err error) {
	reqDoc, err := parseDocument(requestXML)
	if err != nil {
		return nil, nil, err
	}
	respDoc, err := parseDocument(responseXML)
	if err != nil {
		return nil, nil, err
	}
	reqRoot = reqDoc.documentElement()
	respRoot = respDoc.documentElement()
	if reqRoot == nil || respRoot == nil {
		return nil, nil, errNoRootElement
	}
	return reqRoot, respRoot, nil
}

var errNoRootElement = &noRootElementError{}

type noRootElementError struct{}

func (*noRootElementError) Error() string { return "xmlops: document has no root element" }

// firstByLocal / allByLocal search root (inclusive) and all descendants for
// element(s) with the given local tag name, mirroring processor.py's
// _first_by_local/_all_by_local (etree.QName(...).localname comparisons,
// namespace-agnostic).
func firstByLocal(root *xNode, local string) *xNode {
	els := allByLocal(root, local)
	if len(els) == 0 {
		return nil
	}
	return els[0]
}

func allByLocal(root *xNode, local string) []*xNode {
	var out []*xNode
	var walk func(n *xNode)
	walk = func(n *xNode) {
		if n.kind == kindElement && n.local == local {
			out = append(out, n)
		}
		for _, c := range n.children {
			walk(c)
		}
	}
	walk(root)
	return out
}

// deepCopyNode returns a detached deep copy of n and its subtree, mirroring
// processor.py's _deep_copy (copy.deepcopy).
func deepCopyNode(n *xNode) *xNode {
	cp := *n
	cp.parent = nil
	cp.attrs = append([]attrNode{}, n.attrs...)
	cp.nsDecls = append([]nsDecl{}, n.nsDecls...)
	cp.children = make([]*xNode, len(n.children))
	for i, c := range n.children {
		child := deepCopyNode(c)
		child.parent = &cp
		cp.children[i] = child
	}
	return &cp
}

// asLocalTag returns a deep copy of el with its own tag renamed to
// localTag/ns if it doesn't already match, mirroring processor.py's
// _as_local_tag.
func asLocalTag(el *xNode, localTag, ns string) *xNode {
	cp := deepCopyNode(el)
	if cp.local != localTag {
		cp.local = localTag
		cp.prefix = ""
		cp.uri = ns
	}
	return cp
}

// serializeStandalone renders el as a standalone top-level element (with its
// own namespace declaration), matching processor.py's
// etree.tostring(proc, encoding="unicode", xml_declaration=False).
func serializeStandalone(el *xNode) string {
	el.parent = nil
	doc := &xNode{kind: kindDocument, children: []*xNode{el}}
	return string(serializeDocument(doc))
}

func jsonStringArray(items []string) string {
	var b strings.Builder
	b.WriteByte('[')
	for i, s := range items {
		if i > 0 {
			b.WriteByte(',')
		}
		b.WriteByte('"')
		for _, r := range s {
			switch r {
			case '"':
				b.WriteString(`\"`)
			case '\\':
				b.WriteString(`\\`)
			default:
				b.WriteRune(r)
			}
		}
		b.WriteByte('"')
	}
	b.WriteByte(']')
	return b.String()
}
